src/Tasks/ResolveAssemblyReferences/Server/RarController.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Internal;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Contract;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Services;
+using Microsoft.VisualStudio.Threading;
+
+using StreamJsonRpc;
+
+#nullable enable
+namespace Microsoft.Build.Tasks.ResolveAssemblyReferences.Server
+{
+    internal sealed class RarController : IRarController
+    {
+        /// <summary>
+        /// The amount of time to wait for the validation of connection.
+        /// </summary>
+        private const int ValidationTimeout = 60000;
+
+        /// <summary>
+        /// Name of <see cref="NamedPipeServerStream"/>
+        /// </summary>
+        private readonly string _pipeName;
+
+        /// <summary>
+        /// Handshake used for validaion of incoming connections
+        /// </summary>
+        private readonly Handshake _handshake;
+
+
+        /// <summary>
+        /// Factory callback to NamedPipeUtils.CreateNamedPipeServer
+        /// 1. arg: pipe name
+        /// 2. arg: input buffer size
+        /// 3. arg: input buffer size
+        /// 4. arg. number of allow clients
+        /// 5. arg. add right to CreateNewInstance
+        /// </summary>
+        private readonly Func<string, int?, int?, int, bool, NamedPipeServerStream> _namedPipeServerFactory;
+
+        /// <summary>
+        /// Callback to validate the handshake.
+        /// 1. arg: expected handshake
+        /// 2. arg: named pipe over which we should validate the handshake
+        /// 3. arg: timeout for validation
+        /// </summary>
+        private readonly Func<Handshake, NamedPipeServerStream, int, bool> _validateHandshakeCallback;
```

Comment:
Now that you already have two delegates, have you considered passing a object implementing a callback interface (currently with two methods) instead? It may be more readable with parameters named directly in code rather than in comments.

Reply:
That is good idea. I have tried it to make it work, but run into some issues. All of them are related to the availability of Handshake struct in Framework proejct. If I wanted to make it available I would have to link quite a few files. That doesn't seem like a good idea to me. I have tried some other approaches, but none of them seemed reasonable. 

So I decided to leave it like that for now, since in future these callbacks should be removed anyway (with upgrade of TFM of Tasks project, we could link NamedPipeUtil and call the methods directly).